tempto-core/src/main/java/com/teradata/tempto/s3/S3Client.java

```diff
@@ -0,0 +1,232 @@
+            return;
+        }
+
+        List<String> parts = PATH_SPLITTER.splitToList(path);
+        StringBuilder sb = new StringBuilder();
+
+        // Loop through all folders in path and create folders if not already exists
+        for (String objectName : parts) {
+            sb.append(objectName);
+            sb.append(PATH_SEPARATOR);
+            if (!exists(sb.toString())) {
+                createEmptyObject(sb.toString());
+            }
+        }
+    }
+
+    private void createEmptyObject(String path) {
+        // Create meta-data for your folder and set content-length to 0 (which indicates a folder)
+        ObjectMetadata metadata = new ObjectMetadata();
+        metadata.setContentLength(0);
+
+        InputStream emptyContent = new ByteArrayInputStream(new byte[0]);
+        PutObjectRequest putObjectRequest = new PutObjectRequest(bucketName,
+                ensureTrailingSlashPresent(path), emptyContent, metadata);
+
+        // Send request to S3 to create folder
+        client.putObject(putObjectRequest);
+    }
+
+    public void deleteFolder(String path)
+    {
+        List<String> objects = getAllObjects(path);
+
+        // For all objects retrieved, delete by sorted order of furthest away from path first
+        Ordering<String> ordering = Ordering
+                .from(Comparator
+                        .comparingInt(object -> -PATH_SPLITTER.splitToList(object).size()));
+        ordering.sortedCopy(objects).forEach(o -> deleteObject(o));
+
+        // Delete the folder itself
+        deleteObject(ensureTrailingSlashPresent(path));
+    }
+
+    private List<String> getAllObjects(String path)
+    {
+        // Get list of all objects underneath this path
+        ListObjectsV2Result listing = client.listObjectsV2(new ListObjectsV2Request()
+                .withBucketName(bucketName)
+                .withPrefix(ensureTrailingSlashPresent(path)));
+
+        // Return all objects (filtering out the path itself)
+        return listing.getObjectSummaries().stream()
+                .map(S3ObjectSummary::getKey)
+                .filter(key -> !key.equals(ensureTrailingSlashPresent(path)))
+                .collect(Collectors.toList());
+    }
+
+    private void deleteObject(String path)
+    {
+        client.deleteObject(bucketName, path);
+    }
+
+    public void deleteFile(String path)
+    {
+        if (!path.endsWith(PATH_SEPARATOR)) {
+            deleteObject(path);
+        } else {
+            throw new RuntimeException("Folder was passed in to deleteFile: " + path);
+        }
+    }
+
+    public void saveFile(String path, InputStream stream, long byteLength)
+    {
+        ObjectMetadata metadata = new ObjectMetadata();
+        metadata.setContentLength(byteLength);
+
+        client.putObject(new PutObjectRequest(bucketName, path, stream, metadata));
+
+        logger.debug("Saved file {} to S3", path);
+    }
+
+    public OutputStream readFile(String path)
```

Comment:
Modified this to return the InputStream instead of the OutputStream.  I actually thought the same thing that it should return an InputStream but forgot to go back and check on this.